File: src/emulator/core/cartridge/mbc/MBC.cs

namespace DMSharp
{
    public abstract class MBC
    {
        public ExternalBus ext;
        public ushort romBank = 0;
        public static int romBankSize = 16384;
        public static int ramBankSize = 8192;

        public int calcBankAddrRom(ushort addr, ushort bank)
        {
            return (bank * MBC.romBankSize) + (addr - 0x4000);
        }

        public byte readBank(ushort addr, ushort bank)
        {
            bank = (byte)(bank % this.ext.romBanks);
            var calculated = this.calcBankAddrRom(addr, bank);
            return this.ext.rom[calculated];
        }

        public abstract void Reset();
        public abstract byte Read(ushort addr);
        public abstract void Write(ushort addr, byte value);
    }

    public abstract class MBCWithRAM : MBC
    {
        public byte ramBank = 0;
        public bool enableExternalRam = false;
        public byte[] externalRam = new byte[32768];
        public int externalRamDirtyBytes = 0;

        public byte readBankRam(ushort addr, byte bank)
        {
            var calculated = this.calcBankAddrRam(addr, bank);
            return this.externalRam[calculated];
        }

        public void writeBankRam(ushort addr, byte value, byte bank)
        {
            var calculated = this.calcBankAddrRam(addr, bank);
            this.externalRam[calculated % 32768] = value;
            this.externalRamDirtyBytes++;
        }

        public int calcBankAddrRam(ushort addr, byte bank)
        {
            return (bank * MBC.ramBankSize) + (addr - 0xA000);
        }
    }
}

File: src/gameboy/core/memory/mbc/mbc.ts
import MemoryBus from "../../memory/memorybus";
import ExternalBus from "../externalbus";
import { hex } from "../../../tools/util";

export default interface MBC {

    ext: ExternalBus;
    romBank: number;

    read(addr: number): number;
    write(addr: number, value: number): void;

    reset(): void;
}


export default class MBC {
    romBank = 0;
    static romBankSize = 16384;
    static ramBankSize = 8192;

    calcBankAddrRom(addr: number, bank: number): number {
        return (bank * MBC.romBankSize) + (addr - 0x4000);
    }

    readBank(addr: number, bank: number): number {
        bank %= this.ext.romBanks;
        const calculated = this.calcBankAddrRom(addr, bank);
        return this.ext.rom[calculated];
    }
}

export class MBCWithRAM extends MBC {
    ramBank = 0;
    enableExternalRam = false;
    externalRam: Uint8Array = new Uint8Array(32768).fill(0xFF);
    externalRamDirtyBytes = 0;

    readBankRam(addr: number, bank: number): number {
        const calculated = this.calcBankAddrRam(addr, bank);
        return this.externalRam[calculated];
    }

    writeBankRam(addr: number, bank: number, value: number) {
        const calculated = this.calcBankAddrRam(addr, bank);
        this.externalRam[calculated] = value;
        this.externalRamDirtyBytes++;
    }

    calcBankAddrRam(addr: number, bank: number): number {
        return (bank * MBC.ramBankSize) + (addr - 0xA000);
    }
}
